src/group_theory/submonoid/operations.lean

```diff
@@ -512,24 +512,45 @@ namespace monoid_hom
 
 open submonoid
 
-/-- The range of a monoid homomorphism is a submonoid. -/
+/-- For many categories (monoids, modules, rings, ...) the set-theoretic image of a morphism `f` is
+a subobject of the codomain. When this is the case, it is useful to define the range of a morphism
+in such a way that the underlying carrier set of the range subobject is definitionally
+`set.range f`. In particular this means that the types `↥(set.range f)` and `↥f.range` are
+interchangeable without proof obligations.
+
+Sometimes one has a convenient candidate definition for range, which is mathematically correct, but
+which lacks the desired definitional convenience. In such a case one may resort to the `copy`
+pattern. A `copy` function converts the definitional problem for the carrier set of a subobject
+into a one-off propositional proof obligation which one discharges while writing the definition of
+the definitionally convenient range (the parameter `hs` in the example below).
+
+A good example is the case of a morphism of monoids. A convenient definition for `mrange` would be
+`(⊤ : submonoid M).map f`. However since this lacks the required definitional convenience, we
+first define `copy` as follows:
+```lean
+protected def copy (S : submonoid M) (s : set M) (hs : s = S) : submonoid M :=
+{ carrier  := s,
+  one_mem' := hs.symm ▸ S.one_mem',
+  mul_mem' := hs.symm ▸ S.mul_mem' }
+```
+and then finally define:
+```lean
+def mrange (f : M →* N) : submonoid N :=
+((⊤ : submonoid M).map f).copy (set.range f) set.image_univ.symm
+```
+-/
+library_note "range copy pattern"
```

Comment:
@eric-wieser Please feel free to edit my words here liberally. I am attempting to describe your idea!